### PR TITLE
AKS: enable isVnetManaged, add caching

### LINKED ISSUE #2484: AzureManagedControlPlane with existing vnet and existing subnet attempts to delete subnet when cluster is deleted
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**

If you use a `AzureManagedControlPlane` with an existing `virtualNetwork.name` and an existing `virtualNetwork.subnet.name` it will properly use the existing resources. However, when deleting the cluster it will attempt to delete the subnet specified in `virtualNetwork.subnet.name`. You are then stuck and unable to finish the deletion of the cluster without also losing the existing subnet (if it already has resources in it).

**What did you expect to happen:**

While the documentation doesn't specifically say that you can use an existing subnet, I read it as implied because you can use an existing vnet. If that's not accurate, I can open a PR to update the documentation accordingly. However, it would be nice if you could use an existing subnet inside of an existing vnet when creating new AKS clusters.

In our environment, we have a single subnet in our vnets that we'd like to use for new AKS clusters being created with CAPZ as we do a migration so that we don't have to redo any peering connections.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

This seems similar to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1590

**Environment:**

- cluster-api-provider-azure version: registry.k8s.io/cluster-api-azure/cluster-api-azure-controller:v1.4.0
- Kubernetes version: (use `kubectl version`): 1.22.11
- OS (e.g. from `/etc/os-release`): n/a

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind bug
/kind feature

**What this PR does / why we need it**:

This PR enables the `IsVnetManaged()` method for AKS scenarios, which results in actual GET calls against the Azure API to determine in real-time whether the expected capz owner tag (tag is applied one time during cluster create) is present on the Virtual Network.

This will address the subnet delete issue during custom VNET scenarios.

Additionally, in order to account for additional calls against the Azure APIs we've implemented a local cache to ensure we don't call the Azure APIs more than once for VNET information per reconciliation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2484

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
AKS: enable isVnetManaged, add caching
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -276,7 +276,7 @@ create-management-cluster: $(KUSTOMIZE) $(ENVSUBST) $(KUBECTL) $(KIND) ## Create
 	./hack/create-custom-cloud-provider-config.sh
 
 	# Deploy CAPI
-	curl --retry $(CURL_RETRIES) -sSL https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.1.4/cluster-api-components.yaml | $(ENVSUBST) | $(KUBECTL) apply -f -
+	curl --retry $(CURL_RETRIES) -sSL https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.2.0/cluster-api-components.yaml | $(ENVSUBST) | $(KUBECTL) apply -f -
```
Comment (CecileRobertMichon) [inline Makefile:279]: this branch might need to be rebased, not sure why v1.2.0 changes are showing up here

File: azure/services/virtualnetworks/virtualnetworks.go [hunk 2]
```diff
@@ -150,19 +167,34 @@ func (s *Service) IsManaged(ctx context.Context) (bool, error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.IsManaged")
 	defer done()
 
+	// Report errors if the client is rate limited.
+	if !s.RateLimiter.Reader.TryAccept() {
```
Comment (CecileRobertMichon) [inline azure/services/virtualnetworks/virtualnetworks.go:171]: This is good for mitigation when we do run into rate limiting. I agree we should try to do this generically across all services (we're already using the retry-after header to requeue create/delete async calls but we need to also use it across all GETs).

However, there is also an aspect of prevention to avoid running into rate limits as much as possible, by reducing the number of calls we make per reconcile (and reducing how often we reconciling which we already do by debouncing successful reconciles thanks to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/pkg/coalescing/reconciler.go).

Low-hanging fruit for this PR would be to remove any extra calls to isVnetManaged since each one makes an API call. The ideal would be to cache the result of isVnetManaged in scope so we only make a single call per AzureCluster reconcile (ie. we don't need to check if it's managed once in the vnet service, AND for every subnet, we can get the result once and then reuse the value for all services, similar to how we cache VM skus in the machine scope.

File: Makefile (hunk 1 above)
Comment (jackfrancis) [reply Makefile:279]: yuck

File: azure/services/virtualnetworks/virtualnetworks.go (hunk 2 above)
Comment (jackfrancis) [reply azure/services/virtualnetworks/virtualnetworks.go:171]: is "cache" in scope as simple as assigning the API result to the scope object instance? (i.e., a new scope object is instantiated in every reconcile loop)

Comment (CecileRobertMichon) [reply azure/services/virtualnetworks/virtualnetworks.go:171]: could be the API result or even just the boolean value of `isVnetManaged` since that's the only thing we need

Comment (CecileRobertMichon) [reply azure/services/virtualnetworks/virtualnetworks.go:171]: see https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L102 for an example

File: azure/scope/managedcontrolplane.go [hunk 3]
```diff
@@ -328,7 +351,17 @@ func (s *ManagedControlPlaneScope) IsIPv6Enabled() bool {
...
+		return to.Bool(s.cache.isVnetManaged)
+	}
+	ctx := context.Background()
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "scope.ManagedControlPlaneScope.IsVnetManaged")
+	defer done()
+	isManaged, err := virtualnetworks.New(s).IsManaged(ctx)
+	if err != nil {
+		log.Error(err, "Unable to determine if ManagedControlPlaneScope VNET is managed by capz", "AzureManagedCluster", s.ClusterName())
+	}
+	return isManaged
```
Comment (CecileRobertMichon) [inline azure/scope/managedcontrolplane.go:364]: can we also set the cache here if it's unset?

File: azure/services/ratelimiter.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+const (
+	AzureServiceRateLimit            bool    = true
+	AzureServiceRateLimitQPS         float32 = 1.0
+	AzureServiceRateLimitBucket      int     = 1
+	AzureServiceRateLimitQPSWrite    float32 = 1.0
+	AzureServiceRateLimitBucketWrite int     = 1
+)
+
+// NewRateLimiter creates new read and write flowcontrol.RateLimiter from RateLimitConfig.
+func NewRateLimiter(config *RateLimitConfig) (flowcontrol.RateLimiter, flowcontrol.RateLimiter) {
```
Comment (CecileRobertMichon) [inline azure/services/ratelimiter.go:60]: What do you think of moving this to a separate PR? it's a great improvement but I think for purposes of merging this quickly and getting it cherry-picked as a bug fix it would be beneficial to keep this PR as lean as possible. 

It would also be good to implement this across all services and doing a separate PR would also us to do that more easily.

File: azure/scope/managedcontrolplane.go [hunk 5]
```diff
@@ -102,13 +105,33 @@ type ManagedControlPlaneScope struct {
...
 	ManagedMachinePools []ManagedMachinePool
 }
 
+// ManagedControlPlaneCache stores common ManagedControlPlane information so we don't have to hit the API multiple times within the same reconcile loop.
+type ManagedControlPlaneCache struct {
+	isVnetManaged *bool
+}
+
+// InitManagedControlPlaneCache sets cached information about the ManagedControlPlane to be used in the scope.
+func (s *ManagedControlPlaneScope) InitManagedControlPlaneCache(ctx context.Context) error {
```
Comment (CecileRobertMichon) [inline azure/scope/managedcontrolplane.go:122]: I don't know if this init function is necessary given that we're fetching the info in IsVnetManaged() if cache is unset (seems redundant), we could probably get away with either:
a) having the init function set the cache at the beginning of the reconcile and then always return the cache value directly in IsVnetManaged() since cache would always be set, or
b) the first time IsVnetManaged() is called, it sets the value in the cache Just In Time and the rest of the time it gets the existing value from the cache

File: azure/scope/managedcontrolplane.go (hunk 3 above)
Comment (CecileRobertMichon) [reply azure/scope/managedcontrolplane.go:364]: (only if you go with approach b) from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2543/files#r946138219)

File: azure/scope/managedcontrolplane.go [hunk 6]
```diff
@@ -102,13 +105,33 @@ type ManagedControlPlaneScope struct {
...
+	cache          *ManagedControlPlaneCache
 
 	AzureClients
 	Cluster             *clusterv1.Cluster
 	ControlPlane        *infrav1exp.AzureManagedControlPlane
 	ManagedMachinePools []ManagedMachinePool
 }
 
+// ManagedControlPlaneCache stores common ManagedControlPlane information so we don't have to hit the API multiple times within the same reconcile loop.
+type ManagedControlPlaneCache struct {
```
Comment (CecileRobertMichon) [inline azure/scope/managedcontrolplane.go:117]: can we do the same thing in cluster scope?

File: azure/services/ratelimiter.go (hunk 4 above)
Comment (jackfrancis) [reply azure/services/ratelimiter.go:60]: +1

File: azure/scope/managedcontrolplane.go (hunk 5 above)
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:122]: I don't have a strong opinion in the abstract. In my view I'd take the practical approach: are there a non-trivial number of reconciliation contexts that won't actually need these cached values? If so, then definitely we should not init first.

I do have an opinion that we should do caching consistently across service scopes. In which case maybe we punt this as well to a separate workstream (not just vnet)? In my manual tests the impact of these vnet GET calls does not materially decrement the allowed requests limits at a speed necessary to contribute to subscription throttling, so I think we can reasonably land this improvement without necessarily needing to couple the caching optimization.

Comment (CecileRobertMichon) [reply azure/scope/managedcontrolplane.go:122]: > are there a non-trivial number of reconciliation contexts that won't actually need these cached values? If so, then definitely we should not init first.

the answer is no, every single reconcile loop will need at least call isVnetManaged at least once (both reconcile and delete) 

> I do have an opinion that we should do caching consistently across service scopes

agreed, do you have other examples of API GET calls that we need to do across multiple services that you're thinking of? We do have some redundant calls to isVnetManaged (especially one per subnet, which could grow as the number of subnets grows for large clusters with many node pools) so I do think caching is a worthy optimization this time around.

Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:122]: I'm happy to scope the caching to this change (and this will be a good precedent going forward for net new changes that result in API calls).

The only reason I did the init is a basic copy/paste of the machine cache reference:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L102

I think doing these on-demand is the more intuitive approach (simpler as well), so I'll just modify this to do that.

File: azure/services/ratelimiter.go (hunk 4 above)
Comment (mtougeron) [reply azure/services/ratelimiter.go:60]: +1 to splitting the PR

File: azure/scope/managedcontrolplane.go (hunk 3 above)
Comment (mtougeron) [reply azure/scope/managedcontrolplane.go:364]: +1 to being cached instead of making frequent API calls during the reconcile loop.

File: azure/services/ratelimiter.go (hunk 4 above)
Comment (jackfrancis) [reply azure/services/ratelimiter.go:60]: rate limiting will end up here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2572

File: azure/scope/managedcontrolplane.go (hunk 3 above)
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:353]: done

**Comment by mtougeron** [comment]
Not sure if this is relevant, but should the commit message also be changed to be more relevant to what the PR is actually doing now? I'm not sure if that's being too picky in reviews...

**Comment by jackfrancis** [comment]
> Not sure if this is relevant, but should the commit message also be changed to be more relevant to what the PR is actually doing now? I'm not sure if that's being too picky in reviews...

Yes, will definitely change the title!

**Comment by CecileRobertMichon** [comment]
> Yes, will definitely change the title!

can you please also amend the commit? That's what will show up in commit history

Let's also add a release note for this change

File: azure/scope/managedcontrolplane.go (hunk 6 above)
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:121]: done

File: azure/scope/managedcontrolplane.go [hunk 7]
```diff
@@ -328,7 +341,18 @@ func (s *ManagedControlPlaneScope) IsIPv6Enabled() bool {
...
 // IsVnetManaged returns true if the vnet is managed.
 func (s *ManagedControlPlaneScope) IsVnetManaged() bool {
-	return true
+	if s.cache.isVnetManaged != nil {
+		return to.Bool(s.cache.isVnetManaged)
+	}
+	ctx := context.Background()
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "scope.ManagedControlPlaneScope.IsVnetManaged")
+	defer done()
+	isManaged, err := virtualnetworks.New(s).IsManaged(ctx)
```
Comment (mtougeron) [inline azure/scope/managedcontrolplane.go:350]: In `cluster.go` you called this variable `isVnetManaged`.

File: azure/scope/managedcontrolplane.go [hunk 8]
```diff
@@ -82,6 +84,10 @@ func NewManagedControlPlaneScope(ctx context.Context, params ManagedControlPlane
 		}
 	}
 
+	if params.Cache == nil {
+		return nil, errors.New("failed to generate new scope from nil Cache")
```
Comment (mtougeron) [inline azure/scope/managedcontrolplane.go:88]: nit: `failed to generate new scope due to nil Cache`

File: azure/scope/cluster.go [hunk 9]
```diff
@@ -87,6 +88,10 @@ func NewClusterScope(ctx context.Context, params ClusterScopeParams) (*ClusterSc
 		}
 	}
 
+	if params.Cache == nil {
+		return nil, errors.New("failed to generate new scope from nil Cache")
```
Comment (mtougeron) [inline azure/scope/cluster.go:92]: nit: `failed to generate new scope due to nil Cache`

File: azure/services/backoff.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/*
```
Comment (mtougeron) [inline azure/services/backoff.go:1]: I thought the retry handling was going to wait until a followup PR? Or is this just going to be the first place where implemented?

File: azure/services/virtualnetworks/virtualnetworks.go [hunk 11]
```diff
@@ -150,19 +154,29 @@ func (s *Service) IsManaged(ctx context.Context) (bool, error) {
...
 
 	vnetIface, err := s.Get(ctx, spec)
-	if err != nil {
-		return false, err
-	}
 	vnet, ok := vnetIface.(network.VirtualNetwork)
 	if !ok {
 		return false, errors.Errorf("%T is not a network.VirtualNetwork", vnetIface)
 	}
+	if err != nil {
```
Comment (mtougeron) [inline azure/services/virtualnetworks/virtualnetworks.go:171]: Doesn't this error check need to happen before it tries to cast the type into the vnet variable in case it didn't get a response due to rate limiting?

File: azure/scope/cluster.go [hunk 12]
```diff
@@ -87,6 +88,10 @@ func NewClusterScope(ctx context.Context, params ClusterScopeParams) (*ClusterSc
 		}
 	}
 
+	if params.Cache == nil {
```
Comment (CecileRobertMichon) [inline azure/scope/cluster.go:91]: since Cache isn't strictly required, why fail if it's nil? we could simply initialize it to an empty cache if nil (that way no need to add the empty value everywhere in the controllers)

File: azure/services/backoff.go (hunk 10 above)
Comment (CecileRobertMichon) [reply azure/services/backoff.go:1]: +1 let's remove this file

File: azure/services/virtualnetworks/virtualnetworks.go [hunk 13]
```diff
@@ -150,19 +154,29 @@ func (s *Service) IsManaged(ctx context.Context) (bool, error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.IsManaged")
 	defer done()
 
+	// Report errors if the client is throttled.
+	if s.RetryAfter.Reader.After(time.Now()) {
```
Comment (CecileRobertMichon) [inline azure/services/virtualnetworks/virtualnetworks.go:158]: should this also be removed?

Comment (CecileRobertMichon) [reply azure/services/virtualnetworks/virtualnetworks.go:158]: Note: we have a better way to do this with controllers by scheduling a the controller to requeue on the retryAfter time, but let's do that in #2572 and keep this simple for now

File: azure/scope/cluster.go (hunk 12 above)
Comment (jackfrancis) [reply azure/scope/cluster.go:91]: This will work, thanks!

File: azure/scope/managedcontrolplane.go [hunk 14]
```diff
@@ -328,7 +341,18 @@ func (s *ManagedControlPlaneScope) IsIPv6Enabled() bool {
 
 // IsVnetManaged returns true if the vnet is managed.
 func (s *ManagedControlPlaneScope) IsVnetManaged() bool {
-	return true
+	if s.cache.isVnetManaged != nil {
+		return to.Bool(s.cache.isVnetManaged)
+	}
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:347]: It's really a question of trying to limit the refactor scope, but we can definitely include, I agree it would be better.

File: azure/scope/managedcontrolplane.go (hunk 7 above)
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:350]: There's nothing mysterious here, I'm just preferring to use a lowercase local var to capture the response from the method name that we're calling.

File: azure/scope/managedcontrolplane.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply azure/scope/managedcontrolplane.go:347]: If it's going to be too big of a refactor, I'll settle with a TODO and a github issue to pass in a real context

[the remark this answers is not in the document]
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:347]: Yeah, it spans multiple interfaces and is going to be a significant amount of boilerplate change:

image

[the remark this answers is not in the document]
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go:347]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2581

File: azure/services/virtualnetworks/virtualnetworks.go (hunk 11 above)
Comment (jackfrancis) [reply azure/services/virtualnetworks/virtualnetworks.go]: no longer an open question as 429 foo is gone!

File: azure/services/backoff.go (hunk 10 above)
Comment (jackfrancis) [reply azure/services/backoff.go]: done

File: azure/scope/cluster.go (hunk 9 above)
Comment (jackfrancis) [reply azure/scope/cluster.go]: no longer returning error for nil Cache

File: azure/scope/managedcontrolplane.go (hunk 8 above)
Comment (jackfrancis) [reply azure/scope/managedcontrolplane.go]: no longer returning error for nil Cache

File: azure/scope/cluster.go [hunk 15]
```diff
@@ -524,7 +536,12 @@ func (s *ClusterScope) Vnet() *infrav1.VnetSpec {
 
 // IsVnetManaged returns true if the vnet is managed.
 func (s *ClusterScope) IsVnetManaged() bool {
-	return s.Vnet().ID == "" || s.Vnet().Tags.HasOwned(s.ClusterName())
+	if s.cache.isVnetManaged != nil {
+		return to.Bool(s.cache.isVnetManaged)
+	}
+	isVnetManaged := s.Vnet().ID == "" || s.Vnet().Tags.HasOwned(s.ClusterName())
```
Comment (CecileRobertMichon) [inline azure/scope/cluster.go:542]: I wonder if we should change this behavior to be the same as managed clusters eventually, although I agree this is the least dangerous change for now since it preserves existing behavior

Comment (jackfrancis) [reply azure/scope/cluster.go:542]: I was honestly a bit confused when implementing this for clusters, so glad to hear that suggestion.

I can throw up really quick PR after this one lands so that the scope of this PR leans a bit more towards managed clusters only.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
> Additionally, in order to account for additional calls against the Azure APIs I've wired in a rate limiting + HTTP 429 retry solution (inspired by cloud-provider-azure).

Remove this from PR description?

**Comment by jackfrancis** [comment]
> > Additionally, in order to account for additional calls against the Azure APIs I've wired in a rate limiting + HTTP 429 retry solution (inspired by cloud-provider-azure).
> 
> Remove this from PR description?

Done

**Comment by jackfrancis** [comment]
@mtougeron @CecileRobertMichon this is ready for another (hopefully final? 🤞) review round

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Review by mtougeron — APPROVED** [review]
looks good to me.

/approve
